src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs

```diff
@@ -983,6 +983,72 @@ public void Modified_Entry_Adding_Outputs_Records_Added_Step_State()
             });
         }
 
+        [Fact, WorkItem(66451, "https://github.com/dotnet/roslyn/issues/66451")]
+        public void Node_Table_When_Previous_Was_Larger()
+        {
+            ImmutableArray<ImmutableArray<string>> values = ImmutableArray.Create(ImmutableArray.Create("class1"), ImmutableArray.Create("class2"));
+            var inputNode = new InputNode<ImmutableArray<string>>(_ => values).WithTrackingName("Input");
+            var transformNode = new TransformNode<ImmutableArray<string>, string>(inputNode, (arr, ct) => arr[0] == "class3" ? ImmutableArray<string>.Empty : arr, name: "SelectMany");
+
+            DriverStateTable.Builder dstBuilder = GetBuilder(DriverStateTable.Empty, trackIncrementalGeneratorSteps: false);
+            _ = dstBuilder.GetLatestStateTableForNode(transformNode);
```

Comment:
nit: Consider also asserting the table entries from the `transformNode` before making changes